### DISCUSSION TITLE
CustomProxy and best place to catch/intercept error, specifically when JWT token expires.

### DISCUSSION THREAD

**Discussion by richardbartley** [MarimerLLC/csla#3653] [Questions] [answered]
I have a Winforms application calling ASPNet Core hosted DataPortal with a mixed authentication model (AD authentication and good old-fashioned manual logins), and all working fine.

Am using a CustomProxy class, but I need/want a simple re-try pattern if any of my dataportal calls fail with an unauthorized exception when the token expires.

Where's the best place to centralize a check?

I'm looking at the custom proxy class and wondering if I override the base Fetch, Delete, etc methods.

```
public class CustomHttpProxy : HttpProxy
{
    private readonly string _jwtToken;

    public CustomHttpProxy(Csla.ApplicationContext applicationContext, HttpClient httpClient, HttpProxyOptions options, string jwtToken)
        : base(applicationContext, httpClient, options)
    {
        _jwtToken = jwtToken;
    }

    public override Task<DataPortalResult> Fetch(Type objectType, object criteria, DataPortalContext context, bool isSync)
    {
        **// some retry logic here if token expired?????**
        return base.Fetch(objectType, criteria, context, isSync);
    }

```

**Comment by rockfordlhotka** [comment]
I think that is correct.

Rocky

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: richardbartley ***@***.***>
Sent: Monday, January 15, 2024 9:36:30 AM
To: MarimerLLC/csla ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MarimerLLC/csla] CustomProxy and best place to catch/intercept error, specifically when JWT token expires. (Discussion #3653)


I have a Winforms application calling ASPNet Core hosted DataPortal with AD authentication, and all working fine.
Am using a CustomProxy class, but I need/want a simple re-try pattern if any of my dataportal calls fail with an unauthorized exception when the token expires.

Where's the best place to centralize a check?

I'm looking at the custom proxy class and wondering if I override the base Fetch, Delete, etc methods.

public class CustomHttpProxy : HttpProxy
{
    private readonly string _jwtToken;

    public CustomHttpProxy(Csla.ApplicationContext applicationContext, HttpClient httpClient, HttpProxyOptions options, string jwtToken)
        : base(applicationContext, httpClient, options)
    {
        _jwtToken = jwtToken;
    }

    public override Task<DataPortalResult> Fetch(Type objectType, object criteria, DataPortalContext context, bool isSync)
    {
        **// some retry logic here if token expired?????**
        return base.Fetch(objectType, criteria, context, isSync);
    }

**Comment by StefanOssendorf** [comment, marked as answer]
Another way would be to implement a `DelegatingHandler` and configure the the default `HttpClient` (configured via `services.AddHttpClient()` iirc) to use it. So it's automatically used for every `HttpClient` request made by the default `HttpClient`.

**Reply by richardbartley** [reply]
In the end, I found my my scenario was to follow Stefans advice and I configured a policy handler with my desired retry logic.

`services.AddHttpClient<HttpClient>().AddPolicyHandler(GetRetryPolicy());`

The Polly policies can be configured in multiple ways.

For example,

```
private static IAsyncPolicy<HttpResponseMessage> GetRetryPolicy() => HttpPolicyExtensions
     .HandleTransientHttpError()
     .OrResult(msg => msg.StatusCode == System.Net.HttpStatusCode.Unauthorized)
     .RetryAsync(3, async (response, retryCount, context) =>
     {
        // add your retry logic here, like refreshing the token and making 
        // it available on the subsequent http/dataportal request
     });
```
